source/developers-guide/shopware-5-upgrade-guide-for-developers/index.md

```diff
@@ -17,6 +17,374 @@ including minor and bugfix releases, refer to the `UPGRADE.md` file found in you
+    {
+        // Do some payment things
+        $token = $this->get('shopware.components.cart.payment_token')->generate();
+        
+        $returnParameters = [
+            'controller' => 'payment_paypal',
+            'action' => 'return',
+            PaymentTokenService::TYPE_PAYMENT_TOKEN => $token
+        ];
+        $returnLink = $this->router->assemble($returnParameters);
+        
+        $redirectUrl = $this->paymentProviderApi->createPayment($this->getCart(), $returnLink);
+        
+        $this->redirect($redirectUrl);
+    }
+}
+```
+
+### Replaced Codemirror with Ace-Editor
+
+Codemirror has been replaced with the [Ace-Editor](https://ace.c9.io/). For compatibility reason, Ace-Editor supports all xtypes / classes from Codemirror.
+
+The following modes are available:
+- css
+- html
+- javascript
+- json
+- less
+- mysql
+- php
+- sass
+- scss
+- smarty
+- sql
+- text
+- xml
+- xquery
+
+The Ace-Editor has some advantages over the previous editor:
+- It provides syntax validation (see [Improved ExtJs Error Reporter in Backend](#improved-extjs-error-reporter-in-backend))
+- It supports autocompletion
+- It is faster
+
+You can see the autocompletion in action in the mail templates of Shopware 5.6: It autompletes the available Smarty
+variables and tags like `if` or `foreach`. This might be extended to other areas as well in future releases.
+
+If you are interested in implementing this functionality in your own plugin, you'll first have to
+[register a `completer`-callback](https://github.com/shopware/shopware/blob/83b7f50837b134050a8d882cde1dbb3e66c61df9/themes/Backend/ExtJs/backend/mail/view/main/content_editor.js#L107) function.
+This [`callback`](https://github.com/shopware/shopware/blob/83b7f50837b134050a8d882cde1dbb3e66c61df9/themes/Backend/ExtJs/backend/mail/view/main/content_editor.js#L193) determines if an autocompletion could be possible (by performing some [sanity checks](https://github.com/shopware/shopware/blob/83b7f50837b134050a8d882cde1dbb3e66c61df9/themes/Backend/ExtJs/backend/mail/view/main/content_editor.js#L197)) and doing an AJAX-request containing the relevant text portion of the editor. 
+
+The backend now can respond with a data structure containing the relevant autocompletion suggestions. How these are
+to be determined depends on your use case and underlying data structures. The `MailBundle` uses an
+`\Shopware\Bundle\MailBundle\AutoCompleteResolver` that pipes the given text through multiple
+`\Shopware\Bundle\MailBundle\AutocompleteResolver\Resolver`s, each checking for specific possible completions.
+
+### Improved ExtJs Error Reporter in Backend
+
+When an error occurred in the JavaScript of the Backend, the Error Reporter that pops up in these cases often wasn't
+very helpful since the stacktrace being shown in such cases can be overwhelming.
+
+Thanks to the new [Ace-Editor](#replaced-codemirror-with-ace-editor), the Error Reporter can now show you the exact
```

Comment:
?

Reply:
Why thanks to ace? It's the error reporter.

Reply:
It's the Ace editor that does the syntax check and shows the exact line the error occurs